### PR TITLE
Expose the RzCore meta API

### PR DESCRIPTION
**Pull request by b1llow**
# SQUASH ME

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by b1llow** [comment]
https://github.com/rizinorg/cutter/pull/2861

**Comment by XVilka** [comment]
I think analysis should not depend on IO. Better to move those APIs to librz/core/cmeta.c

File: librz/analysis/meta.c [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 #include <rz_analysis.h>
 #include <rz_core.h>
+#include <rz_io.h>
```
Comment (XVilka) [inline librz/analysis/meta.c]: Forgotten?

File: librz/analysis/meson.build [hunk 2]
```diff
@@ -292,7 +292,7 @@ pkgconfig_mod.generate(rz_analysis,
     'rz_diff',
     'rz_bin',
     'rz_flag',
-    'rz_type'
+    'rz_type',
```
Comment (XVilka) [inline librz/analysis/meson.build]: Unnecessary change

File: librz/core/cmeta.c [hunk 3]
```diff
@@ -363,3 +363,109 @@ RZ_IPI void rz_core_meta_editor(RzCore *core, RzAnalysisMetaType mtype, ut64 add
...
+		return false;
+	}
+	*ds = rz_list_first(str_list);
+	rz_list_free(str_list);
+	rz_str_ncpy(*((char **)name), (*ds)->string, limit);
+	(*name)[limit] = '\0';
+	return true;
+}
+
+RZ_API bool rz_core_meta_string_add(RzCore *core, ut64 addr, ut64 size, RzStrEnc encoding, RZ_NULLABLE const char *name) {
```
Comment (XVilka) [inline librz/core/cmeta.c:435]: Please add doxygen

File: librz/core/cmeta.c [hunk 4]
```diff
@@ -363,3 +363,109 @@ RZ_IPI void rz_core_meta_editor(RzCore *core, RzAnalysisMetaType mtype, ut64 add
...
+		result = rz_meta_set_with_subtype(core->analysis, RZ_META_TYPE_STRING, encoding, addr, n, guessname);
+	} else {
+		result = rz_meta_set_with_subtype(core->analysis, RZ_META_TYPE_STRING, encoding, addr, n, name);
+	}
+out:
+	free(guessname);
+	return result;
+}
+
+RZ_API bool rz_core_meta_pascal_string_add(RzCore *core, ut64 addr, RzStrEnc encoding, RZ_NULLABLE const char *name) {
```
Comment (XVilka) [inline librz/core/cmeta.c:477]: Doxygen

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/analysis/meta.c (hunk 1 above)
Comment (b1llow) [reply librz/analysis/meta.c]: Yes i forgot it

**Comment by b1llow** [comment]
doxygen runs much slower than compiled code

File: librz/analysis/meta.c (hunk 1 above)
Comment (b1llow) [reply librz/analysis/meta.c]: I accidentally deleted the `#include <rz_io.h>` of analysis.c, but there seems to be no problem

**Comment by XVilka** [comment]
```
Ubuntu clang-format version 13.0.1-++20211221063253+abc0cf92bb14-1~exp1~20211221183333.39
librz/core/cmeta.c:427:3: error: code should be clang-formatted [-Wclang-format-violations]
 * 
  ^
librz/core/cmeta.c:431:37: error: code should be clang-formatted [-Wclang-format-violations]
 * \param encoding string's encoding 
                                    ^
```

**Comment by b1llow** [comment]
- [ ] `a` asyncCmdEsil * 9, asyncCmd * 1 cmdRawAt * 2 cmdj * 3, cmdjAt * 2
- [ ] `p` cmdRawAt * 2
- [ ] `f` cmdRaw * 8, cmdj * 1
- [ ] `e` cmdRaw * 1, cmdj * 6
- [ ] `w` cmdRaw * 2, cmdj * 1
- [ ] `o` cmdRaw * 4, cmdj * 1
- [ ] `d` asyncCmd * 10, cmdRaw * 2, cmdj * 8
- [ ] `i` cmdj * 11, cmdjTask * 1
- [ ] `v` cmdj * 1
- [x] `C`
- [ ] `z` cmdj * 1
- [ ] `.` cmdTask * 1

https://github.com/rizinorg/rizin/issues/1342
Some of the more used commands have not been porting to the tree-sitter-based one

**Comment by b1llow** [comment]
@XVilka The code of some subprojects does not appear to be formatted

```
git-clang-format --extensions c,cpp,h,hpp,inc --style file dev
changed files:
    librz/core/cmeta.c
    subprojects/rizin-shell-parser/src/parser.c
    subprojects/rizin-shell-parser/src/scanner.c
    subprojects/rizin-shell-parser/src/tree_sitter/parser.h
    subprojects/rzheap/rz_jemalloc/internal/arena.h
    subprojects/rzheap/rz_jemalloc/internal/assert.h
    subprojects/rzheap/rz_jemalloc/internal/atomic.h
    subprojects/rzheap/rz_jemalloc/internal/base.h
    subprojects/rzheap/rz_jemalloc/internal/bitmap.h
    subprojects/rzheap/rz_jemalloc/internal/chunk.h
    subprojects/rzheap/rz_jemalloc/internal/chunk_dss.h
    subprojects/rzheap/rz_jemalloc/internal/chunk_mmap.h
    subprojects/rzheap/rz_jemalloc/internal/ckh.h
    subprojects/rzheap/rz_jemalloc/internal/ctl.h
    subprojects/rzheap/rz_jemalloc/internal/extent.h
    subprojects/rzheap/rz_jemalloc/internal/hash.h
    subprojects/rzheap/rz_jemalloc/internal/huge.h
    subprojects/rzheap/rz_jemalloc/internal/jemalloc_internal.h
    subprojects/rzheap/rz_jemalloc/internal/jemalloc_internal_decls.h
    subprojects/rzheap/rz_jemalloc/internal/jemalloc_internal_defs.h
    subprojects/rzheap/rz_jemalloc/internal/jemalloc_internal_macros.h
    subprojects/rzheap/rz_jemalloc/internal/mb.h
    subprojects/rzheap/rz_jemalloc/internal/mutex.h
    subprojects/rzheap/rz_jemalloc/internal/nstime.h
    subprojects/rzheap/rz_jemalloc/internal/pages.h
    subprojects/rzheap/rz_jemalloc/internal/ph.h
    subprojects/rzheap/rz_jemalloc/internal/private_namespace.h
    subprojects/rzheap/rz_jemalloc/internal/prng.h
    subprojects/rzheap/rz_jemalloc/internal/prof.h
    subprojects/rzheap/rz_jemalloc/internal/public_namespace.h
    subprojects/rzheap/rz_jemalloc/internal/ql.h
    subprojects/rzheap/rz_jemalloc/internal/qr.h
    subprojects/rzheap/rz_jemalloc/internal/quarantine.h
    subprojects/rzheap/rz_jemalloc/internal/rb.h
    subprojects/rzheap/rz_jemalloc/internal/rtree.h
    subprojects/rzheap/rz_jemalloc/internal/size_classes.h
    subprojects/rzheap/rz_jemalloc/internal/smoothstep.h
    subprojects/rzheap/rz_jemalloc/internal/spin.h
    subprojects/rzheap/rz_jemalloc/internal/stats.h
    subprojects/rzheap/rz_jemalloc/internal/tcache.h
    subprojects/rzheap/rz_jemalloc/internal/ticker.h
    subprojects/rzheap/rz_jemalloc/internal/tsd.h
    subprojects/rzheap/rz_jemalloc/internal/util.h
    subprojects/rzheap/rz_jemalloc/internal/valgrind.h
    subprojects/rzheap/rz_jemalloc/internal/witness.h
    subprojects/rzheap/rz_jemalloc/jemalloc.h
    subprojects/rzheap/rz_jemalloc/jemalloc_defs.h
    subprojects/rzheap/rz_jemalloc/jemalloc_macros.h
    subprojects/rzheap/rz_jemalloc/jemalloc_mangle.h
    subprojects/rzheap/rz_jemalloc/jemalloc_mangle_jet.h
    subprojects/rzheap/rz_jemalloc/jemalloc_protos.h
    subprojects/rzheap/rz_jemalloc/jemalloc_protos_jet.h
    subprojects/rzheap/rz_jemalloc/jemalloc_rename.h
    subprojects/rzheap/rz_jemalloc/jemalloc_typedefs.h
    subprojects/rzheap/rz_windows/windows_heap.h
    subprojects/rzspp/bin/main.c
    subprojects/rzspp/config.def.h
    subprojects/rzspp/p/acr.h
    subprojects/rzspp/p/cpp.h
    subprojects/rzspp/p/pod.h
    subprojects/rzspp/p/sh.h
    subprojects/rzspp/p/spp.h
    subprojects/rzspp/rz_api.c
    subprojects/rzspp/rz_api.h
    subprojects/rzspp/spp.c
    subprojects/rzspp/spp.h
```

**Comment by XVilka** [comment]
@imbillow yeah, it's a recent change, this script should be changed:
https://github.com/rizinorg/rizin/blob/dev/sys/clang-format.py#L14
cc @ret2libc

**Review by XVilka — APPROVED** [review]

File: librz/core/cmeta.c [hunk 5]
```diff
@@ -363,3 +363,128 @@ RZ_IPI void rz_core_meta_editor(RzCore *core, RzAnalysisMetaType mtype, ut64 add
 		free(out);
 	}
 }
+
+/**
+ * RzCoreMetaString
+ * @{
```
Comment (ret2libc) [inline librz/core/cmeta.c:369]: ?

File: librz/core/cmeta.c [hunk 6]
```diff
@@ -363,3 +363,128 @@ RZ_IPI void rz_core_meta_editor(RzCore *core, RzAnalysisMetaType mtype, ut64 add
...
+		result = rz_meta_set_with_subtype(core->analysis, RZ_META_TYPE_STRING, encoding, addr, n, name);
+	}
+out:
+	free(guessname);
+	return result;
+}
+
+/**
+ * \brief add a pascal string to RzCore
+ * \see rz_core_meta_string_add
```
Comment (ret2libc) [inline librz/core/cmeta.c]: please just copy the things from the `rz_core_meta_string_add` doxygen doc, otherwise if/when `rz_core_meta_string_add` is removed or the parameters changes we will for sure forget to update these as well.

**Review by ret2libc — COMMENTED** [review]
LGTM apart from that small thing

File: librz/core/cmeta.c (hunk 5 above)
Comment (XVilka) [reply librz/core/cmeta.c:369]: It's a grouping Doxygen feature.

Comment (XVilka) [reply librz/core/cmeta.c:369]: https://www.doxygen.nl/manual/grouping.html

Comment (ret2libc) [reply librz/core/cmeta.c:369]: woah, ok :D

**Review by XVilka — APPROVED** [review]
